### DISCUSSION TITLE
WalkForward split periods in calendar units, like Weeks, Months,

### DISCUSSION THREAD

**Discussion by jmlocatelli** [skfolio/skfolio#41] [Ideas]
Hi, for the investor perpective, it would be nice to have the possibility to split and synchronize the train and test periods for the rolling portfolio as calendar unit, like weeks, months and others. 
Ex.:
WalkForward(train_size=12, test_size=3, unit="Month")

**Comment by HugoDelatte** [comment]
The difficulty is that the cross-validators have no information about data frequency (the input X could  be a numpy array, a DataFrame with no datatime index etc.). So it needs to be done by the user:
* WalkForward(train_size=12\*21, test_size=3\*21) if the returns are daily
* WalkForward(train_size=12, test_size=3) if the returns are monthly
*  etc

Would that work for your use-case?

**Reply by jmlocatelli** [reply]
Not really. I would define as: - having daily prices, rebalance (test) the portfolio every 2 calendar months with a lookback (train) period of 1 calendar year. Using 252 (trading) days to represent a year is not precise for the investor perspective. Month would start always at the first trading day of the month. Year would always start in the first trading day of the year.
It would be a much complex splitter, and yes, it would need datetime index.

**Reply by HugoDelatte** [reply]
I think that's a task for a backtesting package. Cross-validators in `skfolio` (as well as scikit-learn) are used for model validation, selection and parameter tuning. Usually, what's important when choosing a model, is the rebalancing frequency and training size, but not the exact index of that training set (otherwise you risk to overfit the model). Final step would be to backtest the strategy based on the skfolio model with a backtesting package (including the exact rebalancing dates) . 

However, if really needed, it's still possible to use `skfolio` as a basic backtesting tool by creating custom cross-validators.

**Reply by jmlocatelli** [reply]
I was thinking the same. Backtesting is another animal. I'm also "playing" with vectorbr and pyportfolioopt. skfolio is feature reach for model validation, selection and parameter tuning, but lacks some bt functionality. vectorbt is reach in bt functionaly but poor in terms of modeling features. Mixing them could be a good way. I normally use Zorro and Amibroker for rapid strategy tests, but they are not so good for portfolio optimization. Zorro is very simple and good for bt, and has a python interface. I may use Zorro for trade control and skfolio for modeling. It's not the ideal solution for me. I like to keep all process as simple as possible. But I'm not going to give it up so easy. And thank you very much for your kindly support.

**Reply by SamoPP** [reply]
I almost opened an issue and accidentally clicked on discussions and saw this thread... :)

There are a bunch of calendar effects/seasonality in  real world trading that we are exploiting. And ideally one would like to see what happens if rebalancing "schedule" is changed and observe and asses the reality of calendar effects.

For example I often want to test the effect of more or less frequent portfolio rebalancing. I want to see what is "better": rebalance weekly, monthly, quarterly, yearly. But, trading day of the week/month/quarter when the rebalancing is made has a material influence on the result.

Or, for example, one would say, nah, calendar effects are an illusion and you are just overoptimizing recent past (like half a century of turn of the month for example). But, if you believe this turn of the month effects (or day of the week effects) are an illusion, then a prudent thing is to do some "tranching" of the portfolio and create a (equal weight) portfolio of portfolios that each rebalance on different days of the week or trading day of the month.

As I understand there is no way to do this currently in skfolio. I would love to see some of this functionality in skfolio. Any chance to add it?

**Reply by HugoDelatte** [reply]
I think you are right, it would be interesting to test that effect. 
Let's brainstorm on the best API to achieve this.

I like the idea of @jmlocatelli : `WalkForward(train_size=12, test_size=3, unit="Month")`

I would rename `unit` by `period` to be homogenous to pandas naming and use an `Enum` named `Period` that enumerates [all pandas periods](https://pandas.pydata.org/docs/user_guide/timeseries.html#timeseries-period-aliases).

We also need to specify when that period start: the `n` calendar  day of the period. We could call it `day_start_period` (feel free to suggest a better  name).

For example, that would give:
* `WalkForward(train_size=12,  test_size=3, period=Period.Monthly, day_start_period=1)` : Rebalance every 3 months starting the first calendar day of the month (or the next following date that is part of the `X` datetime index), and train on the preceding 12 months.
*  `WalkForward(train_size=52,  test_size=3, period=Period.Weekly, day_start_period=5)`: Rebalance every 3 weeks starting the fifth calendar day of the week, so Friday (or the next following date that is part of the `X` datetime index), and train on the preceding 52 weeks.

Default would be `period=None` and `day_start_period=None`. When provided, we would perform a check on `X` and raise an error
if it doesn't have an index of type `pd.DatetimeIndex`.

Let me know if that solution would work / propose potential improvements or different approaches.

**Comment by CarloNicolini** [comment]
> However, if really needed, it's still possible to use `skfolio` as a basic backtesting tool by creating custom cross-validators.

Btw, there is a backtester, tested against the matlab one in my [scikit-portfolio develop branch](https://github.com/scikit-portfolio/scikit-portfolio/blob/develop/skportfolio/backtest/backtester.py) it is basic but works.

**Reply by jmlocatelli** [reply]
Good to know. I'll give a look. tks

**Reply by jmlocatelli** [reply]
Did a quick look at the package. Very nice work.
"scikit-portfolio offers a method for backtesting a long-only investment strategy based on the usage of the portfolio estimators shipped with it."
Any plans to work with short investment strategy too?

**Comment by HugoDelatte** [comment]
@jmlocatelli this is now implemented in `WalkForward` where you can split on specific dates using `freq` and `freq_offset` params: https://skfolio.org/generated/skfolio.model_selection.WalkForward.html

**Comment by jmlocatelli** [comment]
Hello Hugo.

Nice to hear from you.

Very good. I appreciate the effort to increase skfolio functionality.

 

Rgds, 

Jose Marcio
